driver-core/src/main/com/mongodb/client/model/expressions/MqlExpressionCodecProvider.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.model.expressions;
+
+import com.mongodb.annotations.Immutable;
+import org.bson.codecs.Codec;
+import org.bson.codecs.configuration.CodecProvider;
+import org.bson.codecs.configuration.CodecRegistry;
+
+@Immutable
+public class MqlExpressionCodecProvider implements CodecProvider {
```

Comment:
MqlExpressionCodec is in the same package as MqlExpression to ensure package-private on `toBsonValue`, which I think is more important to protect than MqlExpressionCodec. So I do not think this can be non-public or in another package.

Reply:
If we move `MqlExpression`, `MqlExpressionCodec `, `MqlExpressionCodecProvider` to `com.mongodb.internal.client.model.expressions`, then they will all be excluded from the public API by means of being in the `internal` package, and `toBsonValue` will stay package-access (though the latter will not be of the essence anymore).

Reply:
`Expressions` might end up using of package-private functionality on the aforementioned. Once this is merged, I will create a draft PR to merge `expressions` into `master`, and I think we should move comments like this there.

Reply:
Discussed in Zoom, and realized that `MqlExpressionCodecProvider` has to be part of public API, because if a user wants to specify a custom codec provider that does not have all the codecs from `MongoClient.getDefaultCodecRegistry`, but also wants to be able to use expressions via the existing methods that accept `Object`(a.k.a. `TExpression`), then a users needs access to `MqlExpressionCodecProvider`<sup>1</sup>.

---

<sup>1</sup> Technically, a user may use `MongoClient.getDefaultCodecRegistry` in its custom codec provider in such a way that it is only used for encoding expressions, but it's a very unfriendly approach. Moreover, if `MqlExpressionCodecProvider` is not exposed, a user don't even have a reliable way of knowing that expressions are encoded via the codecs functionality and that `MongoClient.getDefaultCodecRegistry` contains a non-public codec for them.

Reply:
Also, if we are making `MqlExpressionCodecProvider` part of the public API, we need to document that it is included in `MongoClientSettings.getDefaultCodecRegistry`.